### PR TITLE
implement gunzipbase64

### LINKED ISSUE #800: terraform function gunzipbase64 as opposite of base64gzip is missing
### OpenTofu Version

```shell
v1.6.0-dev
```

### Use Cases

write base64gzip content to any storage, read it and decode it with gunzipbase64

### Attempted Solutions

none available

### Proposal

implement gunzipbase64

### References

https://github.com/hashicorp/terraform/issues/22568

### PR DESCRIPTION
**Pull request by l-with**
add terraform function gunzipbase64 as opposite of base64gzip similar to

- base64decode opposite of base64encode
- textdecodebase64 opposite of textencodebase64
- jsondecode opposite of jsonencode
- yamldecode opposite of yamlencode

Resolves #800

## Target Release

1.6.0

### NEW FEATURES | UPGRADE NOTES | ENHANCEMENTS | BUG FIXES | EXPERIMENTS

 

terraform function gunzipbase64 can be used as opposite of base64gzip, e.g.

- write base64gzip content to any storage, read it and decode it with gunzipbase64

### CODE REVIEW

**Comment by l-with** [comment]
> Reminder for the PR assignee: If this is a user-visible change, please update the changelog as part of the PR.

created #800

File: internal/lang/funcs/encoding.go [hunk 1]
```diff
@@ -178,6 +179,37 @@ var Base64GzipFunc = function.New(&function.Spec{
...
+		str, strMarks := args[0].Unmark()
+		s := str.AsString()
+		sDec, err := base64.StdEncoding.DecodeString(s)
+		if err != nil {
+			return cty.UnknownVal(cty.String), fmt.Errorf("failed to decode base64 data %s", redactIfSensitive(s, strMarks))
+		}
+		sDecBuffer := bytes.NewReader(sDec)
+		gzipReader, err := gzip.NewReader(sDecBuffer)
+		if err != nil {
+			return cty.UnknownVal(cty.String), fmt.Errorf("failed to gunzip base64 decoded data: %w", err)
```
Comment (kislerdm) [inline internal/lang/funcs/encoding.go:202]: **(nitpick)** The data are no longer base64-encoded, hence no need to mention that in the error message IMO.

```suggestion
                        return cty.UnknownVal(cty.String), fmt.Errorf("failed to gunzip bytestream: %w", err)
```

Comment (l-with) [reply internal/lang/funcs/encoding.go]: done

**Comment by kislerdm** [comment]
Hey! Thanks for opening PR, but please follow to the [contribution guideline](https://github.com/opentofu/opentofu/blob/main/CONTRIBUTING.md#contributing-a-code-change) moving forward:

> Only issues with the `accepted` label have been officially accepted for implementation, so please avoid working on issues without that label.

Thanks!

**Comment by kislerdm** [comment]
The development for #800 has been frozen until the stable tofu version is released. Please see the [comment](https://github.com/opentofu/opentofu/issues/800#issuecomment-1798474945) for details.

Thank you!

File: internal/lang/funcs/descriptions.go [hunk 2]
```diff
@@ -53,6 +53,10 @@ var DescriptionList = map[string]descriptionEntry{
 		Description:      "`base64gzip` compresses a string with gzip and then encodes the result in Base64 encoding.",
 		ParamDescription: []string{""},
 	},
+	"gunzipbase64": {
+		Description:      "`gunzipbase64` decodes a Base64 encoded string and uncompresses the result with gzip.",
```
[the remark this answers is not in the document]
Comment (l-with) [reply internal/lang/funcs/descriptions.go]: https://github.com/opentofu/opentofu/pull/799/commits/1504f21b4b36c2d1f53d13d423337cb4740110eb

File: internal/lang/funcs/encoding.go [hunk 3]
```diff
@@ -228,6 +260,13 @@ func Base64Gzip(str cty.Value) (cty.Value, error) {
 	return Base64GzipFunc.Call([]cty.Value{str})
 }
 
+// GunzipBase64 decodes a Base64 encoded string and uncompresses the result with gzip.
+//
+// Terraform uses the "standard" Base64 alphabet as defined in RFC 4648 section 4.
```
[the remark this answers is not in the document]
Comment (l-with) [reply internal/lang/funcs/encoding.go]: https://github.com/opentofu/opentofu/pull/799/commits/db8227892c36cb1a85cdcc5340604fa1c116fc82

File: internal/lang/funcs/encoding.go [hunk 4]
```diff
@@ -228,6 +260,13 @@ func Base64Gzip(str cty.Value) (cty.Value, error) {
 	return Base64GzipFunc.Call([]cty.Value{str})
 }
 
+// GunzipBase64 decodes a Base64 encoded string and uncompresses the result with gzip.
```
[the remark this answers is not in the document]
Comment (l-with) [reply internal/lang/funcs/encoding.go]: https://github.com/opentofu/opentofu/pull/799/commits/11906404b87ee651b5b37e5ca74650d41610f610

File: website/docs/language/functions/gunzipbase64.mdx [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+page_title: gunzipbase64 - Functions - Configuration Language
+description: |-
+  The gunzipbase64 funtion decodes a Base64 encoded string and uncompresses the result with gzip.
+---
+
+# `gunzipbase64` Function
+
+`gunzipbase64` decodes a Base64 encoded string and uncompresses the result with gzip.
+
+Terraform uses the "standard" Base64 alphabet as defined in
```
[the remark this answers is not in the document]
Comment (l-with) [reply website/docs/language/functions/gunzipbase64.mdx]: https://github.com/opentofu/opentofu/pull/799/commits/d50f34e172ef7864a4467c6a84ce03dce127d0f2

File: website/docs/language/functions/gunzipbase64.mdx [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+[RFC 4648 section 4](https://tools.ietf.org/html/rfc4648#section-4).
+
+## Related Functions
+
+* [`base64gzip`](/terraform/language/functions/base64gzip) compresses a string with gzip and then encodes the result in
```
[the remark this answers is not in the document]
Comment (l-with) [reply website/docs/language/functions/gunzipbase64.mdx]: https://github.com/opentofu/opentofu/pull/799/commits/70bf6f9ff55317ee6194bf4a8572d58afc53e5c7

**Comment by l-with** [comment]
> Hey @l-with I haven't tested this change yet, a few language-related nitpicks and docs fixes for now.

The change was implemented test-driven (without granular commits), please have a look at the implemented tests.

**Comment by cube2222** [comment]
Hey, could you please rebase this @l-with?

**Comment by l-with** [comment]
> Hey, could you please rebase this @l-with?

I hope that the sync github suggested me and resulted in the merge is sufficient.

**Comment by cube2222** [comment]
@l-with it's still a mess unfortunately, if you look at the PR, it contains 48 commits. In order to fix this up, you'll need to rebase via git and only include the commits you need. To do that, you need to run from your branch
```
git rebase -i 38eb199282d2d068967c2e01ae3b812279217c73
```
(38eb199282d2d068967c2e01ae3b812279217c73 is the current main branch head commit sha)

and then next to each line where there is a commit that is **not** one of yours, change the beginning from "pick" to "drop". I just tried it, and this was what it ended up being once I was done with it (and worked): https://gist.github.com/cube2222/8c360724d5e6ec0b3c435e01f15cb238

Once you do that, you can force-push, and it should be good.

**Comment by cam72cam** [comment]
Let me know if you would like help with this @l-with

**Comment by l-with** [comment]
reopen

File: website/data/language-nav-data.json [hunk 7]
```diff
@@ -763,8 +767,13 @@
         "hidden": true
       },
       {
-        "title": "base64sha256",
-        "path": "language/functions/base64sha256",
+        "title": "base64gzip",
```
Comment (cam72cam) [inline website/data/language-nav-data.json:770]: It looks like you accidentally added an additional base64gzip entry and removed the base64sha256 entry here.

Comment (cam72cam) [reply website/data/language-nav-data.json:770]: It also looks like this list is alphabetical and your entry for gunzipbase64 should be lower down.

**Comment by cam72cam** [comment]
@l-with The code looks great, only one thing to address in the function documentation listing.

File: website/data/language-nav-data.json (hunk 7 above)
Comment (cam72cam) [reply website/data/language-nav-data.json:770]: @l-with sorry to nit-pick, but base64sha256 should come after base64gzip.  Once that is addressed, this is ready to merge.

**Comment by cam72cam** [comment]
Sorry to add more hassle, but it appears that when you fixed your branch the signatures you added were dropped.

`git rebase --signoff main` should re-add the signatures to your commits.

**Review by cam72cam — APPROVED** [review]

**Comment by cam72cam** [comment]
Thanks @l-with !

**Pull request merged**